src/util/requests/RequestBuilder.h

```diff
@@ -0,0 +1,155 @@
+#include <boost/beast/core/error.hpp>
+#include <boost/beast/http/field.hpp>
+#include <boost/beast/http/message.hpp>
+#include <boost/beast/http/string_body.hpp>
+#include <boost/beast/http/verb.hpp>
+
+#include <chrono>
+#include <string>
+#include <string_view>
+#include <vector>
+
+namespace util::requests {
+
+/**
+ * @brief Builder for HTTP requests
+ */
+class RequestBuilder {
+    std::string host_;
+    std::string port_;
+    std::chrono::milliseconds timeout_{DEFAULT_TIMEOUT};
+    boost::beast::http::request<boost::beast::http::string_body> request_;
+    bool sslEnabled_{false};
+
+public:
+    /**
+     * @brief Construct a new Request Builder object
+     *
+     * @param host host to connect to
+     * @param port port to connect to
+     */
+    RequestBuilder(std::string host, std::string port);
+
+    /**
+     * @brief Add a header to the request
+     *
+     * @param header header to add
+     * @return reference to itself
+     */
+    RequestBuilder&
+    addHeader(HttpHeader const& header);
+
+    /**
+     * @brief Add headers to the request
+     *
+     * @param headers headers to add
+     * @return reference to itself
+     */
+    RequestBuilder&
+    addHeaders(std::vector<HttpHeader> const& headers);
+
+    /**
+     * @brief Add body or data to the request
+     *
+     * @param data data to add
+     * @return reference to itself
+     */
+    RequestBuilder&
+    addData(std::string data);
+
+    /**
+     * @brief Set the timeout for the request
+     *
+     * @note Default timeout is defined in DEFAULT_TIMEOUT
+     *
+     * @param timeout timeout to set
+     * @return reference to itself
+     */
+    RequestBuilder&
+    setTimeout(std::chrono::milliseconds timeout);
+
+    /**
+     * @brief Set the target for the request
+     *
+     * @note Default target is "/"
+     *
+     * @param target target to set
+     * @return reference to itself
+     */
+    RequestBuilder&
+    setTarget(std::string_view target);
+
+    /**
+     * @brief Set SSL enabled or disabled
+     *
+     * @note Default is false
+     *
+     * @param ssl boolean value to set
+     * @return reference to itself
+     */
+    RequestBuilder&
+    setSslEnabled(bool enabled);
+
+    /**
+     * @brief Perform a GET request asynchronously
+     *
+     * @note It is not thread-safe to call get() and post() of the same RequestBuilder from multiple threads. But it is
+     * fine to call only get() or only post() of the same RequestBuilder from multiple threads.
+     *
+     * @param yield yield context
+     * @return expected response or error
+     */
+    Expected<std::string, RequestError>
+    get(boost::asio::yield_context yield);
+
+    /**
+     * @brief Perform a POST request asynchronously
+     *
+     * @note It is not thread-safe to call get() and post() of the same RequestBuilder from multiple threads. But it is
+     * fine to call only get() or only post() of the same RequestBuilder from multiple threads.
```

Comment:
curious why it is okay to call get() or post() in multiple threads but not get() and post()?

Reply:
Because inside get() or post() we modify underling request setting method for it. So the same modifications from different threads are fine, but setting method get and post from different threads at the same time is a data race.